### PR TITLE
feat(replay): Add network breadcrumbs

### PR DESCRIPTION
**Pull request by krystofwoldrich**
## :loudspeaker: Type of change

- [x] New feature

## :scroll: Description

This PR adds network breadcrumbs to RN replays.

## :green_heart: How did you test it?
RN sample app, Android works, ~iOS waiting for https://github.com/getsentry/sentry-react-native/pull/3846~, iOS working

## :pencil: Checklist

- [x] I reviewed submitted code
- [ ] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled
- [x] All tests passing
- [x] No breaking changes

Screenshot 2024-06-24 at 16 36 58

Screenshot 2024-06-25 at 10 59 14

### CODE REVIEW

File: src/js/replay/networkUtils.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  if (RN_GLOBAL_OBJ.TextEncoder) {
+    return new RN_GLOBAL_OBJ.TextEncoder().encode(input);
+  }
+  return utf8ToBytes(input);
+}
+
+function _serializeFormData(formData: FormData): string {
+  // This is a bit simplified, but gives us a decent estimate
+  // This converts e.g. { name: 'Anne Smith', age: 13 } to 'name=Anne+Smith&age=13'
+  // @ts-expect-error passing FormData to URLSearchParams actually works
```
Comment (vaind) [inline src/js/replay/networkUtils.ts:62]:
```suggestion
  // @ts-expect-error passing FormData to URLSearchParams won't correctly serialize `File` entries, which is fine for this use-case. See https://github.com/microsoft/TypeScript/issues/30584
```

**Review by vaind — APPROVED** [review]
LGTM but there are some conflicts with what I've done today in the base PR.

File: test/replay/networkUtils.test.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      expect(getBodySize('abcd')).toBe(4);
+      // Emojis are correctly counted as mutliple characters
+      expect(getBodySize('With emoji: 😈')).toBe(16);
+    });
+
+    it('works with URLSearchParams', () => {
+      const params = new URLSearchParams();
+      params.append('name', 'Jane');
+      params.append('age', '42');
+      params.append('emoji', '😈');
```
Comment (lucas-zimerman) [inline test/replay/networkUtils.test.ts:33]: Q: Should we also test with non english characters?

File: src/js/replay/networkUtils.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import { RN_GLOBAL_OBJ } from '../utils/worldwide';
+import { utf8ToBytes } from '../vendor';
+
+/** Convert a Content-Length header to number/undefined.  */
```
Comment (lucas-zimerman) [inline src/js/replay/networkUtils.ts:4]: NIT
```suggestion
/** Convert a Content-Length header to number/undefined. */
```

**Review by lucas-zimerman — APPROVED** [review]
LGTM!

File: src/js/replay/networkUtils.ts (hunk 1 above)
Comment (krystofwoldrich) [reply src/js/replay/networkUtils.ts:62]: Sounds like something we should add also to the JS repo where I copied this from.

`won't correctly serialize `File` entries, which is fine for this use-case. See https://github.com/microsoft/TypeScript/issues/30584`

File: test/replay/networkUtils.test.ts (hunk 2 above)
Comment (krystofwoldrich) [reply test/replay/networkUtils.test.ts:33]: Probably not necessary, the main goal is to check that the function doesn't just could string length.

**Pull request merged**